sql/sql_acl.cc

```diff
@@ -13196,11 +13257,75 @@ bool Sql_cmd_grant_table::execute_table_mask(THD *thd)
   }
 
   WSREP_TO_ISOLATION_BEGIN(WSREP_MYSQL_DB, NULL, NULL);
-  /* Conditionally writes to binlog */
-  if (mysql_grant(thd, m_gp.db().str, lex->users_list,
-                  {m_gp.object_privilege(), is_revoke()},
-                  false/*not proxy*/))
-    return true;
+  if (m_gp.db().length)
+  {
+    char tmp_db[SAFE_NAME_LEN+1];
+    const char *db= m_gp.db().str;
+    /* Convert database name to lower case if lower_case_table_names. */
+    if (lower_case_table_names)
+    {
+      char *end= strnmov(tmp_db, db, sizeof(tmp_db));
+      if (end >= tmp_db + sizeof(tmp_db))
+      {
+        my_error(ER_WRONG_DB_NAME ,MYF(0), db);
+        return true;
+      }
+      my_casedn_str(files_charset_info, tmp_db);
+      db=tmp_db;
+    }
+
+    /* Only allow DB level grants. */
+    if ((m_gp.object_privilege() & DB_ACLS) != m_gp.object_privilege())
+    {
+        my_error(ER_WRONG_USAGE, MYF(0), "DB GRANT", "GLOBAL PRIVILEGES");
+        return true;
+    }
+
+    if (tables.open_and_lock(thd, Table_user | Table_db, TL_WRITE))
+      return true;
```

Comment:
here, again, you return an error for return value of 1. And also you've lost an assert
```c
   DBUG_ASSERT(!thd->is_current_stmt_binlog_format_row());
```

Reply:
`Sql_cmd_grant_sp::execute` becomes quite different now, it doesn't assert, doesn't binlog, doesn't open tables. Unlike other `Sql_cmd_grant*::execute` methods